ext/curl/multi.c

```diff
@@ -322,96 +297,54 @@ PHP_FUNCTION(curl_multi_info_read)
 
 	/* find the original easy curl handle */
 	{
-		zval	*pz_ch = _php_curl_multi_find_easy_handle(mh, tmp_msg->easy_handle);
+		zval *pz_ch = _php_curl_multi_find_easy_handle(mh, tmp_msg->easy_handle);
 		if (pz_ch != NULL) {
-			/* we are adding a reference to the underlying php_curl
-			   resource, so we need to add one to the resource's refcount
-			   in order to ensure it doesn't get destroyed when the
-			   underlying curl easy handle goes out of scope.
-			   Normally you would call zval_copy_ctor( pz_ch ), or
-			   SEPARATE_ZVAL, but those create new zvals, which is already
-			   being done in add_assoc_resource */
-			Z_ADDREF_P(pz_ch);
-
 			/* we must save result to be able to read error message */
-			ch = (php_curl*)zend_fetch_resource(Z_RES_P(pz_ch), le_curl_name, le_curl);
+			ch = Z_CURL_P(pz_ch);
 			SAVE_CURL_ERROR(ch, tmp_msg->data.result);
 
-			/* add_assoc_resource automatically creates a new zval to
-			   wrap the "resource" represented by the current pz_ch */
-
+			Z_ADDREF_P(pz_ch);
 			add_assoc_zval(return_value, "handle", pz_ch);
 		}
 	}
 }
 /* }}} */
 
-/* {{{ proto void curl_multi_close(resource mh)
+/* {{{ proto void curl_multi_close(CurlMulti mh)
    Close a set of cURL handles */
 PHP_FUNCTION(curl_multi_close)
 {
-	zval      *z_mh;
 	php_curlm *mh;
+	zval *z_mh;
+
+	zend_llist_position pos;
+	zval *pz_ch;
 
 	ZEND_PARSE_PARAMETERS_START(1,1)
-		Z_PARAM_RESOURCE(z_mh)
+		Z_PARAM_OBJECT_OF_CLASS(z_mh, curl_multi_ce)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if ((mh = (php_curlm *)zend_fetch_resource(Z_RES_P(z_mh), le_curl_multi_handle_name, le_curl_multi_handle)) == NULL) {
-		RETURN_THROWS();
-	}
-
-	zend_list_close(Z_RES_P(z_mh));
-}
-/* }}} */
-
-void _php_curl_multi_close(zend_resource *rsrc) /* {{{ */
-{
-	php_curlm *mh = (php_curlm *)rsrc->ptr;
-	if (mh) {
-		zend_llist_position pos;
-		php_curl *ch;
-		zval	*pz_ch;
-
-		for (pz_ch = (zval *)zend_llist_get_first_ex(&mh->easyh, &pos); pz_ch;
-			pz_ch = (zval *)zend_llist_get_next_ex(&mh->easyh, &pos)) {
-			/* ptr is NULL means it already be freed */
-			if (Z_RES_P(pz_ch)->ptr) {
-				if ((ch = (php_curl *) zend_fetch_resource(Z_RES_P(pz_ch), le_curl_name, le_curl))) {
-					_php_curl_verify_handlers(ch, 0);
-				}
-			}
-		}
+	mh = Z_CURL_MULTI_P(z_mh);
 
-		curl_multi_cleanup(mh->multi);
-		zend_llist_clean(&mh->easyh);
-		if (mh->handlers->server_push) {
-			zval_ptr_dtor(&mh->handlers->server_push->func_name);
-			efree(mh->handlers->server_push);
-		}
-		if (mh->handlers) {
-			efree(mh->handlers);
-		}
-		efree(mh);
-		rsrc->ptr = NULL;
+	for (pz_ch = (zval *)zend_llist_get_first_ex(&mh->easyh, &pos); pz_ch;
+		pz_ch = (zval *)zend_llist_get_next_ex(&mh->easyh, &pos)) {
+		zend_llist_del_element(&mh->easyh, pz_ch, (int (*)(void *, void *))curl_compare_objects);
```

Comment:
I think what I had in mind here is to also perform a curl_multi_remove_handle call.